instrumentation/finagle/http/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/finagle/ChannelTransportInstrumentation.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.finagle;
+
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.twitter.util.Future;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import java.lang.invoke.MethodHandle;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+public class ChannelTransportInstrumentation implements TypeInstrumentation {
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return named("com.twitter.finagle.netty4.transport.ChannelTransport");
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        isMethod().and(named("write")),
+        ChannelTransportInstrumentation.class.getName() + "$WriteAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class WriteAdvice {
+
+    @Advice.OnMethodEnter(suppress = Throwable.class, skipOn = Advice.OnDefaultValue.class)
+    public static boolean methodEnter() {
+      // only call when we've effectively wrapped the method (hence: isRecursed)
+      return Helpers.WRITE_GUARD.isRecursed();
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void methodExit(
+        @Advice.SelfCallHandle MethodHandle handle,
+        @Advice.Return(readOnly = false) Future<?> ret) {
+      ret = Helpers.callWrite(handle, ret);
```

Comment:
I don't get why you are doing it like this. If you are going to use the method handle to call this method again shouldn't you do it in the enter advice so you could skip the body for the original execution. Currently the future returned isn't always used as far as I can tell.

Reply:
> If you are going to use the method handle to call this method again shouldn't you do it in the enter advice so you could skip the body for the original execution

Can you explain what you mean by this? In this case we're extracting an object from finagle's Context and wrapping the call to the instrumented method so we can modify the returned value. I'm going to try and make this work a little differently bc to your point I have no idea why I'm doing it like this apart from it was the only way I could get it to, but I'm curious how you mean. Very possible I'm missing something.

Reply:
Thank you for the suggestion to look more closely at this. I revamped this considerably and removed a bunch of code paths and indirection. Personally, I like the result. PTAL when you get a chance :pray: